lib/core/src/display/navigation/navigator.rs

```diff
@@ -0,0 +1,90 @@
+use super::camera_manager::CameraManager;
+use super::camera_manager::Event;
+use super::camera_manager::ZoomingEvent;
+use super::camera_manager::PanningEvent;
+use crate::system::web::Result;
+use crate::display::rendering::Camera;
+use crate::display::rendering::CameraType;
+use crate::display::rendering::DOMContainer;
+
+use nalgebra::Vector3;
+use nalgebra::Vector2;
+
+// =================
+// === Navigator ===
+// =================
+
+/// Navigator enables camera navigation with mouse interactions on the specified DOM.
+pub struct Navigator {
+    dom        : DOMContainer,
+    events     : CameraManager,
+    position   : Vector3<f32>,
+    zoom_speed : f32
+}
+
+impl Navigator {
+    pub fn new(dom:&DOMContainer, position:Vector3<f32>, zoom_speed:f32) -> Result<Self> {
+        let events     = CameraManager::new(dom)?;
+        let dom        = dom.clone();
+
+        Ok(Self { dom, events, position, zoom_speed })
+    }
+
+    /// Navigates the camera and returns it's position.
+    pub fn navigate(&mut self, camera:&mut Camera) -> Vector3<f32> {
+        while let Some(event) = self.events.poll() {
+            match event {
+                Event::Zooming(zoom) => self.zoom(camera, zoom),
+                Event::Panning(pan)   => self.pan(camera, pan)
+            }
+        }
+        self.position
+    }
+
+    fn zoom(&mut self, camera:&mut Camera, zoom: ZoomingEvent) {
+        if let CameraType::Perspective(persp) = camera.camera_type() {
+            let point      = zoom.focus;
+            let normalized = normalize_point2(point, self.dom.dimensions());
+            let normalized = normalized_to_range2(normalized, -1.0, 1.0);
+
+            // Scale X and Y to compensate aspect and fov.
+            let z =  zoom.amount * self.zoom_speed;
+            let x = -normalized.x * z / persp.aspect;
+            let y =  normalized.y * z / camera.get_y_scale();
+
+            self.position += Vector3::new(x, y, z);
+            if  self.position.z < persp.near + 1.0 {
+                self.position.z = persp.near + 1.0;
+            }
+        }
+    }
```

Comment:
This is a generic zooming utility. Why it is in a struct which forces us to use DOM elements?

Reply:
Because it needs to know in which point of the DOM it will be zoomed.